### ISSUE TITLE
Keyboard CW transmission?

### ISSUE THREAD

**Issue by Haasoperator** [df8oe/mchf-github#1496]
I was wondering if any progress has been made on allowing CW to be transmitted via USB keyboard, now that RTTY and BPSK have been enabled. Mainly thinking of mcHF use.

**Comment by WD8BXS** [comment]
This would be a great enhancement, several people have expressed interest to me about it.

**Comment by WD8BXS** [comment]
Is this going to be possible??

**Comment by m-chichikalov** [comment]
Sure, sooner or later, I can take a look into this a bit later.  73!

**Comment by m-chichikalov** [comment]
@WD8BXS , Hi,
Hi, have you ever seen that when you are typing really fast (kind of fast ;), some characters doubles?  For example, if quickly press "qwe" on the screen it appears as -> "qwwe" and will transmit as on the screen.

Just want to know, maybe this is the issue with my keyboard.
73!

BTW: the keyboard CW transmission is almost done.

**Comment by WD8BXS** [comment]
Hi,No I have not seen this issue.I will experiment.Good news on the CW..Thanks for all of your efforts.
Chuck

**Comment by m-chichikalov** [comment]
@WD8BXS , it should be the same on your side, theoretically. Could you try, please, in RTTY or BPSK press 'w' and by holding it press 'e' -> should appear as 'ww'  No?

@db4ple , @df8oe, @Haasoperator ,  could you guys try also, thanks. 

Seems like keyboard sends a buffer of pressed buttons, but we handle only first in the array...

**Comment by m-chichikalov** [comment]
![](https://s8.hostingkartinok.com/uploads/images/2019/01/09152a5a67019e94662b8192b362bed6.jpg)

Ok, there are two ways to handle it, the first is the easiest one, the second one is the right one... 
1 - send requests to keyboard more rarely, -> **tested** -> Doesn't help.
2 - dedicate module to handle keyboards requests, and analyze the entire array...

maybe another?

**Comment by m-chichikalov** [comment]
**FIXED**

**Comment by db4ple** [comment]
Fixed in PR #1693

**Comment by m-chichikalov** [comment]
> Fixed in PR #1693

Not really in #1693. I did not push it yet.

**Comment by db4ple** [comment]
Sorry, yes, I mixed it up with CW entry for PSK.

**Comment by WD8BXS** [comment]
Yes it is true, but I don't type that fast hihi.

So as I understand it, the cw from keyboard is NOT quite ready yet??

TNX, 73, Chuck WD8BXS

**Comment by m-chichikalov** [comment]
> Yes it is true, but I don't type that fast hihi.
> 
> So as I understand it, the cw from keyboard is NOT quite ready yet??
> 
> TNX, 73, Chuck WD8BXS

You are right, it's not pushed to the official build. It's working, but because everything very coupled together, some changes in one place broke some in others.... 

It's working already but needs some fix for RTTY. What do you using, I can compile for you to try it on and give me feedback, it's also very important. 

I added:
- there are three buttons on the UI, so the context of them could be transmitted by pressing F5,F6 and F7 respectively.  
- You can clean the buffer by pressing ESC on the keyboard.
- Fixed issue with pressing multi buttons on the keyboard.
- The context of those "Three buttons" could be written from the keyboard, not only from paddles.

73! Max RV9YW

**Comment by WD8BXS** [comment]
Sound like it is coming along.
I use the MCHF
A friend is interested in the CW from Keyboard on his MCHF.

TNX, Chuck

**Comment by WD8BXS** [comment]
Been playing around, I can't seem to get my radios to sync.
very different frequencies on my MCHF to my IC-7610 to copy each other on RTTY or PSK.

Not sure what is going on there.
I tried with translation on and off, still no sync!!
73,
de Chuck

**Comment by m-chichikalov** [comment]
@WD8BXS , looks like USB keyboard for CW is already in the build, you can try...

**Comment by WD8BXS** [comment]
m-chichikalov

Yes it does work with my wired USB keyboard, but now my Rii Mini X1 Wireless keyboard does not now!!
The F1 and F2 buttons do indeed put it into transmit and receive, but no letters work.
Thanks,
 Chuck

**Comment by m-chichikalov** [comment]
During testing, I also found that sometimes my keyboard not recognized by mcHF. So, I usually fixed it by rebooting TRX. 

> m-chichikalov
> 
> Yes it does work with my wired USB keyboard, but now my Rii Mini X1 Wireless keyboard does not now!!
> The F1 and F2 buttons do indeed put it into transmit and receive, but no letters work.
> Thanks,
> Chuck

Hi Chuck,
During testing, I also found that sometimes my keyboard not recognized by mcHF. So, I usually fixed it by rebooting TRX.  Later will try understand why it's like this...

What CW mode have you used? It would work only with the IAM mode.

**Comment by WD8BXS** [comment]
I tried the reboot, but to no avail.  as I said it must see the wireless keyboard as the F1 and F2 work.

I am using IAM mode.

**Comment by m-chichikalov** [comment]
Ok... I will take a look what we can do to handle different keyboards with different layouts...
However, try another one just in case. Thanks 73!

**Comment by WD8BXS** [comment]
Testing tonight!! it works but only for a minute or so. I can type in only a few words then it stops, and stays keyed. when I get it unkeyed, the keyboard stops working.

Very good progress though.

I know you will get it going ..

Thanks,
 Chuck WD8BXS

**Comment by WD8BXS** [comment]
I just discovered that on 5 watts it works great.. maybe some RF getting into it on 10 watts ??

Chuck

**Comment by m-chichikalov** [comment]
> I just discovered that on 5 watts it works great.. maybe some RF getting into it on 10 watts ??
> 
> Chuck

Hi Chuck, sure... High level of vswr should not help, yes maybe Your antenna is fine-tuned. Don't know, but it's a hardware issue that keyboard stop working after 10Wt, It's up to you how to fix it, sorry We cannot help you with that.

**Comment by Harry-ei6ku** [comment]
Can I ask pls, is there a list of USB keyboards know to work with MCHF, cabled or bluetooth.

73
Harry

**Comment by df8oe** [comment]
Such a list does not exist (and cannot exist because of the huge amount of different keyboards). All standard keyboards (which are no "multiple USB keyboards") should work.

**Comment by m-chichikalov** [comment]
At least We can start creating such a list...
@ei8hvb  The Dell SK-8115 works fine on my side

**Comment by ok1cdj** [comment]
I made page in wiki, you can add your keyboard. Actually i am looking for small keyboard which will works.
https://github.com/df8oe/UHSDR/wiki/Compatible-USB-keyboards
